### PR TITLE
Navigation: Fix creating Navigation from pages or menu with HTML in title

### LINKED ISSUE #23919: Inline styling markup saved as link text in navigation screen
**Describe the bug**

When building a menu with the navigation screen and styling navigation items with the Bold and Italics toolbar buttons, the styling markup generated is being saved as text, so that when adding the menu to a Navigation block in a regular post, `<em>` and `<strong>` tags appear as part of the link text.

**To reproduce**

1. Create a menu using the navigation screen, and add some items.
2. Edit the text of those items to add bold or italics styles, using the block toolbar.
3. Save navigation, and open a new post.
4. Add a Navigation block and choose the menu created above to populate it.
5. Verify that `<em>` and `<strong>` tags are now part of the link text.
**Expected behavior**

Expected not to have markup tags as part of link text.

**Screenshots**

Screen Shot 2020-07-14 at 3 39 35 pm

Screen Shot 2020-07-14 at 1 40 15 pm

**Editor version (please complete the following information):**

Latest master branch from this repo.

### PR DESCRIPTION
**Pull request by noisysocks**
Fixes https://github.com/WordPress/gutenberg/issues/23919.

Pages and menu items in WordPress allow HTML in the title. We should therefore not escape the menu item title when creating a Navigation block from top-level pages or an existing menu.

Additionally, the menu items REST API endpoint was setting `title.raw` to the menu item CPT's `post_title` (`$menu_item->post_title`) and `title.rendered` to the referenced object's `post_title` (`get_post( $menu_item->object_id )->post_title)`). Both `title.raw` and `title.rendered` should be set to the menu item's `title` (`$menu_item->title`), which [is properly set by `wp_setup_nav_menu_item()`](https://github.com/WordPress/wordpress-develop/blob/master/src/wp-includes/nav-menu.php#L836-L853).

The only difference is that `title.rendered` should be passed through `the_title`, [which escapes `&` characters](https://github.com/WordPress/wordpress-develop/blob/ca5c4cebefe9f493e42eba9ec909b433373bf3be/src/wp-includes/default-filters.php#L168-L170), and `nav_menu_item_title`, which provides plugins a chance to customise menu item titles before display. This matches [what `Walker_Nav_Menu` does](https://github.com/WordPress/wordpress-develop/blob/ca5c4cebefe9f493e42eba9ec909b433373bf3be/src/wp-includes/class-walker-nav-menu.php#L212-L225).

### How to test

1. Create a menu which has a menu item with the title: `<strong>Menu item</strong> & foo`
1. Create a page with the title: `<strong>Page</strong> & foo`.
1. Insert a Navigation block, select the menu you just created, and click _Create_.
1. Insert another Navigation  block, select _Create from all top-level pages_, and click _Create_.

Both blocks should have bold text and a `&` in one of the links.

### CODE REVIEW

**Comment by TimothyBJacobs** [comment]
This looks right to me @noisysocks, can we get a phpunit test covering this?

**Comment by noisysocks** [comment]
> This looks right to me @noisysocks, can we get a phpunit test covering this?

Done!

**Comment by noisysocks** [comment]
Huh. The test fails in GitHub Actions but succeeds locally. I'll investigate tomorrow.

File: phpunit/class-rest-nav-menu-items-controller-test.php [hunk 1]
```diff
@@ -210,13 +210,29 @@ public function test_get_item_escapes_title() {
...
 		$response = rest_get_server()->dispatch( $request );
 
-		$this->assertEquals(
-			array(
-				'rendered' => '<strong>Foo</strong> &#038; bar',
-				'raw'      => '<strong>Foo</strong> & bar',
-			),
-			$response->get_data()['title']
-		);
+		if ( ! MULTISITE ) {
```
Comment (TimothyBJacobs) [inline phpunit/class-rest-nav-menu-items-controller-test.php:213]:
```suggestion
		if ( ! is_multisite() ) {
```

**Comment by noisysocks** [comment]
> Huh. The test fails in GitHub Actions but succeeds locally. I'll investigate tomorrow.

Turns out the regular tests were passing but the multisite tests were failing. This is because administrators in a multisite do not have `unfiltered_html`. I updated the test.

File: phpunit/class-rest-nav-menu-items-controller-test.php (hunk 1 above)
Comment (noisysocks) [reply phpunit/class-rest-nav-menu-items-controller-test.php]: 👍

**Review by TimothyBJacobs — APPROVED** [review]

**Pull request merged**